staging/cse/windows/nvidiagpudriverfunc.ps1

```diff
@@ -0,0 +1,244 @@
+        if ($IsSignatureValid -eq $false) {
+            $ErrorMsg = "Signature embedded in $($SetupTarget) is not valid."
+            Write-ConsoleLog $ErrorMsg
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+        }
+        else {
+            Write-ConsoleLog "Signature embedded in $($SetupTarget) is valid."
+        }
+
+  
+        Write-ConsoleLog "Installing $SetupTarget ..."
+        try {
+            $InstallLogFolder = "$LogFolder\NvidiaInstallLog"
+            $Arguments = "-s -n -log:$InstallLogFolder -loglevel:6"
+      
+            $p = Start-Process -FilePath $SetupTarget -ArgumentList $Arguments -PassThru
+  
+            $Timeout = 10 * 60 # in seconds. 10 minutes for timeout of the installation
+            Wait-Process -InputObject $p -Timeout $Timeout -ErrorAction Stop
+      
+            # check if installation was successful
+            if ($p.ExitCode -eq 0 -or $p.ExitCode -eq 1) {
+                # 1 is issued when reboot is required after success
+                Write-ConsoleLog "GPU Driver Installation Success. Code: $($p.ExitCode)"
+            }
+            else {
+                $ErrorMsg = "GPU Driver Installation Failed! Code: $($p.ExitCode)"
+                Write-ConsoleLog $ErrorMsg
+                Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+            }
+  
+            if ($Setup.RebootNeeded -or $p.ExitCode -eq 1) {
+                Write-ConsoleLog "Reboot is needed for this GPU Driver..."
+                $RebootNeeded = $true
+            }
+            return $RebootNeeded
+        }
+        catch [System.TimeoutException] {
+            $ErrorMsg = "Timeout $Timeout s exceeded. Stopping the installation process. Reboot for another attempt."
+            Write-ConsoleLog $ErrorMsg
+            Stop-Process -InputObject $p -Force
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_TIMEOUT -ErrorMessage $ErrorMsg
+        }
+        catch {
+            $ErrorMsg = "Exception: $($_.ToString())"
+            Write-ConsoleLog $ErrorMsg # the status file may get over-written when the agent re-attempts this step
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+        }
+      
+    }
+    catch {
+        $FatalError += $_
+        $errorCount = $FatalError.Count
+        $ErrorMsg = "A fatal error occurred. Number of errors: $errorCount. Error details: $($_ | Out-String)"
+        Write-ConsoleLog $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_FAILED -ErrorMessage $ErrorMsg
+    }
+}
+  
+function Get-Setup {
+    param(
+        [Parameter(Mandatory = $true)]
+        [PSCustomObject]$DriverUrlConfig
+    )
+
+    [OutputType([hashtable])]
+    
+    $GpuDriverURL = $DriverUrlConfig.GpuDriverURL
```

Comment:
what is the endpoint of the url? We need to assure that they can be accessed in different clouds with custom's firewall settings.

Reply:
The url can be set differently for different clouds. Currently in the rp PR draft, they are all set to be the same urls. Those urls don't work on the private clouds but we have the flexibility to configure them.
RP PR: [https://dev.azure.com/msazure/CloudNativeCompute/_git/aks-rp/pullrequest/8878419?_a=files&path=/resourceprovider/server/microsoft.com/datamodel/model/aksengine_azenvtypes.go](https://github.com/Azure/AgentBaker/pull/url)

Chris and I are currently working on creating a pipeline to release the drivers to acs mirror. Then we will change the url on the rp side to make sure they work on private clouds.